### PR TITLE
feat: Introduce TOTP recipe

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Introduce TOTP recipe

## Related issues

-   https://github.com/supertokens/supertokens-python/issues/324
-   https://github.com/supertokens/supertokens-core/issues/554

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.
 
## Remaining TODOs for this PR

-   [ ] Implement the classes
-   [ ] Add tests

### CODE REVIEW

File: supertokens_python/recipe/totp/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    pass
+
+
+class UnknownDeviceError:
+    pass
+
+
+class CreateDeviceOkResult:
+    def __init__(self, user_identifier: str, issuer_name: str, secret: str):
+        self.qr = f"otpauth://totp/{quote(issuer_name)}:{quote(user_identifier)}?secret={secret}&issuer={quote(issuer_name)}"
```
Comment (KShivendu) [inline supertokens_python/recipe/totp/interfaces.py:44]: Is this okay. Or should I just store constructor args here?

File: supertokens_python/recipe/totp/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+if TYPE_CHECKING:
+    from .utils import OverrideConfig as InputOverrideConfig
+
+
+def init(
+    issuer: str,
+    skew: Optional[int] = None,
+    period: Optional[int] = None,
+    allow_unverified_devices: Optional[bool] = False,
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/__init__.py:26]: what is the use case of making allow_unverified_devices True?

File: supertokens_python/recipe/totp/asyncio/__init__.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    device_name: str,
+    skew: Optional[int] = None,
+    period: Optional[int] = None,
+    user_context: Optional[Dict[str, Any]] = None,
+) -> Union[CreateDeviceOkResult, DeviceAlreadyExistsError]:
+    ...
+
+
+async def verify_code(
+    user_id: str, code: int, user_context: Optional[Dict[str, Any]] = None
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/asyncio/__init__.py:43]: are you sure code is an int? The CDI spec has it as type of string

File: supertokens_python/recipe/totp/asyncio/__init__.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+) -> Union[
+    VerifyDeviceOkResult, InvalidTotpError, TotpNotEnabledError, LimitReachedError
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/asyncio/__init__.py:45]: TotpNotEnabledError -> does this mean the license key is not put? Cause that results in a 4xx error to be thrown from the core.

File: supertokens_python/recipe/totp/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+
+class TotpNotEnabledErrorResponse(APIResponse):
+    status = "TOTP_NOT_ENABLED_ERROR"
+
+    def to_json(self):
+        return {"status": self.status}
+
+
+class LimitReachedErrorResponse(APIResponse):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/interfaces.py:142]: Should contain retryAfterMs?

File: supertokens_python/recipe/totp/asyncio/__init__.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    ...
+
+
+async def update_device(
+    user_id: str,
+    existing_device_name: str,
+    new_device_name: str,
+    user_context: Optional[Dict[str, Any]] = None,
+) -> Union[
+    VerifyDeviceOkResult, InvalidTotpError, TotpNotEnabledError, UnknownDeviceError
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/asyncio/__init__.py:67]: Output type is just wrong.

File: supertokens_python/recipe/totp/asyncio/__init__.py (hunk 4 above)
Comment (rishabhpoddar) [reply supertokens_python/recipe/totp/asyncio/__init__.py:45]: If it does not mean the license key is missing, then when is this error thrown?

File: supertokens_python/recipe/totp/interfaces.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+class InvalidTotpError:
+    pass
+
+
+class TotpNotEnabledError:
+    pass
+
+
+class LimitReachedError:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/interfaces.py:34]: retryAfterMs

Comment (rishabhpoddar) [reply supertokens_python/recipe/totp/interfaces.py:34]: Please make sure to not loose out on info from the CDI

File: supertokens_python/recipe/totp/interfaces.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    @abstractmethod
+    async def update_device_put(
+        self, user_id: str, existing_device_name: str, new_device_name: str
+    ) -> Union[
+        UpdateDevicePutOkResponse,
+        TotpNotEnabledErrorResponse,
+        DeviceAlreadyExistsErrorResponse,
+        UnknownDeviceErrorResponse,
+    ]:
+        pass
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/interfaces.py:256]: we don't need this API (update_device_put)

File: supertokens_python/recipe/totp/interfaces.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    status = "OK"
+
+    def __init__(self, devices: List[Dict[str, Any]]):
+        self.devices = devices
+
+    def to_json(self):
+        return {"status": self.status, "devices": self.devices}
+
+
+class APIInterface:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/interfaces.py:210]: missing API to get the secret stuff to disable on the frontend

File: supertokens_python/recipe/totp/interfaces.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    def __init__(self) -> None:
+        self.disable_create_device_post = False
+        self.disable_verify_code_post = False
+        self.disable_verify_device_post = False
+        self.disable_update_device_put = False
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/interfaces.py:215]:
```suggestion
```

File: supertokens_python/recipe/totp/interfaces.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        self.disable_remove_device_post = False
+        self.disable_list_devices_get = False
+
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/totp/interfaces.py:218]: all these APIs should not take a user_id. They should take a session container instead.

**Pull request closed without merging**